controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/DeploymentQuotaCalculator.java

```diff
@@ -0,0 +1,63 @@
+package com.yahoo.vespa.hosted.controller.application;
+
+import com.yahoo.config.application.api.DeploymentSpec;
+import com.yahoo.config.provision.ApplicationId;
+import com.yahoo.config.provision.zone.ZoneId;
+import com.yahoo.vespa.hosted.controller.Application;
+import com.yahoo.vespa.hosted.controller.api.integration.billing.Quota;
+
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+import java.util.List;
+
+/** Calculates the quota to allocate to a deployment. */
+public class DeploymentQuotaCalculator {
+
+    public static Quota calculate(Quota tenantQuota,
+                                  List<Application> tenantApps,
+                                  ApplicationId deployingApp, ZoneId deployingZone,
+                                  DeploymentSpec deploymentSpec) {
+
+        if (tenantQuota.budget().isEmpty()) return tenantQuota; // Shortcut if there is no budget limit to care about.
+
+        if (deployingZone.environment().isProduction()) return probablyEnoughForAll(tenantQuota, tenantApps, deployingApp, deploymentSpec);
+
+        return getMaximumAllowedQuota(tenantQuota, tenantApps, deployingApp, deployingZone);
+    }
+
+    /** Just get the maximum quota we are allowed to use. */
+    private static Quota getMaximumAllowedQuota(Quota tenantQuota, List<Application> applications,
+                                                ApplicationId application, ZoneId zone) {
+        var usageOutsideDeployment = applications.stream()
+                .map(app -> app.quotaUsage(application, zone))
```

Comment:
An ApplicationId is not the identifier of an Application - that is the TenantAndApplicationId (which it knows).

An ApplicationId actually specifies an instance: so TenantAndApplicationId + InstanceName. I could have passed just the InstanceName, and in the quotaUsage methods reconstructed the ApplicationId - but it would be more code and didn't seem any clearer to me.

ApplicationId needs to be renamed to InstanceId, or perhaps TenantAndApplicationAndInstanceId, but I think that's a big refactor :-)

Reply:
Aha! That makes more sense.